oeps/oep-0045/decisions/0001-tutor-as-replacement-for-edx-configuration.rst

```diff
@@ -0,0 +1,328 @@
+Tutor as a replacement for ``edx/configuration``
+================================================
+
+
+Status
+------
+
+Draft
+
+
+Context
+-------
+
+With the acceptance of OEP-0045, it becomes clear that `edx/configuration <https://github.com/edx/configuration/>`__
+will be deprecated and that a replacement is required.  However, OEP-0045 does not go into the details of what such an
+alternative would look like.
+
+This ADR identifies what criteria any replacement should meet, evaluates the `Tutor
+<https://github.com/overhangio/tutor>`__ project against it, outlining work that needs to be done, and proposes the
+latter as the choice of the community.
+
+
+Criteria
+--------
+
+Any replacement for ``edx/configuration`` must provide the following:
+
+1.  A full, scalable deployment strategy for edx-platform and its main IDAs, supporting both small and large production
+    deployments
+2.  A mechanism that is suitable for automated deployment and testing
+3.  Support for running the latest so-called "open release" of Open edX, and a method to upgrade from the previous one
+4.  Support for running the master branch not only of edx-platform, but its corresponding IDAs
+5.  Support for customizing of the installation, from themes to edx-platform forks, including installation of custom
+    requirements
+6.  A documented way to migrate existing deployments from the native installation
+
+And importantly:
+
+7.  The project must be healthy, with thoughtful technical leadership, a well-organized, maintainable codebase, and an
+    active community
+
+
+Evaluation
+----------
+
+It is under the light of the previously described `criteria <#criteria>`__ that the Tutor project is evaluated.
+
+
+Criterion 1. How well can Tutor deploy Open edX for production use?
+...................................................................
+
+
+Single server
+`````````````
+
+
+Tutor's stated purpose is "To make it possible to deploy, administer and upgrade Open edX anywhere, easily." (`source
+<https://docs.tutor.overhang.io/faq.html#what-is-the-purpose-of-tutor>`__).  And for single server installs, it clearly
+lives up to it: in terms of ease-of-deployment, it is a step up from ``edx/configuration``. Getting a server up is
+simpler than even using the official Docker devstack. A local test Koa deployment had one up and running in `under 15
```

Comment:
Devstack is, first and foremost, edX's development environment. I don't see a clear path towards the unification of the Tutor dev environment and the Devstack, other than convincing edX developers that Tutor is superior in terms of developer experience.

Reply:
This is an important point that we may want to address specifically in either this or a subsequent ADR.  It is definitely a separate can of worms, though.  A development environment has some needs that are not part of a production one.  Such as, for instance, automatic code and asset reloading.  I haven't evaluated Tutor with that in mind - yet.

Reply:
Do we have a detailed list of the things Tutor doesn't do that it would need to do to be a development environment?

Reply:
Yup, and devstack isn't just edX' development environment, it's also the one used by many contributors & providers. If we can all agree on a common development enviroment, that will help pull our resources together, get more contributions focused on one project, and everyone will have less maintenance to do, including edX. @regisb this is a good opportunity to actually do that convincing, while we are looking at Tutor. :)

Reply:
Tutor was designed from the ground up to work _both_ as a production and a [development](https://docs.tutor.overhang.io/dev.html) environment. These days, when I need to work on the named releases I prefer to use Tutor than the devstack -- I might be biased, but I am a core contributor on both projects, so I think I have a balanced opinion.

I'm not sure how I can preach in favor of Tutor as a better development environment, but I can try to list some of the arguments:

1. Easier and lighter Docker image building and pulling
2. Easier extensibility with plugins, such that it's easier to launch a minimal dev environment
3. Better and faster static asset reloading
4. The [machine](https://overhang.io/tutor/plugin/machine) plugin which I recently released and now can't live without

For the record, my marketing strategy never was to pretend that Tutor was in any way superior to the devstack or the native installation. Evangelization is not my thing. People choose Tutor because it's a tool that works for them. I'm happy to show people how they can make the best use of Tutor but I'm not sure how else I can convince them.

Reply:
@regisb, I think it was just a question of me not having evaluated Tutor for suitability as a devstack - it certainly looks good from what you say.  I'm going to take a specific look, this sprint - I had already set aside some time to do so - and include the results here in the ADR.

Thanks for already listing the pros, though! :+1:

Reply:
@regisb That sounds look good arguments to me :+1: In addition to @arbrandes testing this sprint (thanks!), @nasthagiri mentioned wanting to start a discussion about this - it sounds like great next steps, by getting all people involved to understand the differences, and see what decision makes sense.

Reply:
If people want to test Tutor as a devstack, please get in touch to get a free Wizard Edition license, so that you can try out the machine plugin: https://overhang.io/tutor/plugin/machine I think it's a game-changer for Open edX development.
(or even better, you can show your support by buying a license :stuck_out_tongue:)

Also, I am working on running edx-platform unit tests onm Tutor and got pretty good results: only 35 tests are failing (out of 17k) with pytest. I suspect most of those are also failing on the devstack when they are run with pytest. The Docker test image will be published soon, along with instructions on how to run unit tests.

Reply:
Exciting!  Cc: @jmbowman @jinder1s @feanil See Regis's note above as we prep this exploratory effort by the Arch team.

Reply:
As promissed, I added a "Devstack Criterion" to the ADR and evaluated Tutor against it.  See the latest commit
(e06cc82).  Spoiler: the setup that more closely mirrors the official devstack is a tad more involved than the latter, but it works well!

(@regisb, does the advertised "machine plugin" make it a bit easier? Didn't have time to test it.  I did [send over a PR](https://github.com/overhangio/tutor/pull/397) to fix a runtime issue I faced, in any case.)

Cc @nasthagiri - now that I did my homework, I'm ready to help organize that devstack meeting.  More on this tomorrow at the contributor's meeting.

Reply:
> does the advertised "machine plugin" make it a bit easier?

It does not make the process easier, but it makes you run Docker/docker-compose on a remote server (like [Appsembler's Sultan](https://github.com/appsembler/sultan)) so image re-building and requirement downloading is much faster.

> Is it open source?

yes, it is AGPLv3 but it is not free: https://overhang.io/tutor/plugin/machine

Could we clarify that a prerequisite should be that Tutor be a good development environment, but it is not a prerequisite that Tutor replaces the devstack?